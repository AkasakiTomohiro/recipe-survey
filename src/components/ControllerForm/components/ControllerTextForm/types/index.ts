/**
 * @file index.ts
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import { FieldPath } from 'react-hook-form';

import { IGridFormControlProps } from '@/components/ControllerForm';
import { RequiredFilterType } from '@/types';
import { TextFieldProps } from '@mui/material/TextField';

export type IControllerTextFormProps<T extends Record<string, any> = Record<string, any>> = IGridFormControlProps & Omit<TextFieldProps, "variant" | "error"> & {

  /**
   * フォームの名前
   */
   label: string;

   /**
    * controlで管理しているオブジェクトのキーを指定
    */
   name: FieldPath<RequiredFilterType<T, number | string>>;

   /**
    * デフォルトで表示したいヘルパーテキスト
    *
    * @details バリデーションエラーがある場合は表示されない
    */
   inputErrorHelperText?: string;

   /**
    * 編集可能かどうか
    */
   isReadOnly?: boolean;
}