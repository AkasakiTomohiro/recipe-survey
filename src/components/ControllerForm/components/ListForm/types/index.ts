/**
 * @file index.ts
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

export interface IListFormProps {

  /**
   * フォーム
   */
  forms: (index: number) => JSX.Element[];

  /**
   * controlで管理しているオブジェクトのキーを指定
   */
  name: string;

  /**
   * 配列の要素を増やすときの初期値
   */
  initialValue: object;

  /**
   * 最大要素数
   */
  maxLength: number;

  /**
   * 編集可能かどうか
   * @default
   */
  editable?: boolean;
}

export type FormRowProps = {
  index: number;
  length: number;
  forms: (index: number) => JSX.Element[];
  onClickDelete: (index: number) => void;
  editable?: boolean;
}