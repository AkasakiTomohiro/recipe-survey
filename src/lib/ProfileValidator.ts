/**
 * @file ProfileValidator.ts
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import * as z from 'zod';

export const ProfileValidator = z.object({
  profileInfo: z.object({

    /**
     * プロファイルID
     */
    id: z.string().uuid(),

    /**
     * プロファイル名
     */
    name: z.string().min(1).max(20)
  })
});
export type Profile = z.infer<typeof ProfileValidator>;

export const MaterialTypesValidator = z
  .object({

    /**
     * 素材種別
     */
    materialTypes: z.array(z.object({
      materialType: z.string().min(1).max(10)
    })).min(1).max(10)
  });
export type MaterialTypes = z.infer<typeof MaterialTypesValidator>;

export const UnitsValidator = z
  .object({
    units: z.array(z.object({

      /**
       * 単位
       */
      unit: z.string().min(1).max(20),

      /**
       * 説明
       */
      description: z.string().min(1).max(128)

    })).min(1).max(10)
  });

export type Units = z.infer<typeof UnitsValidator>;

export type ProfileSet = MaterialTypes & Profile & Units;

export const profileSetValidator = ProfileValidator.and(MaterialTypesValidator).and(UnitsValidator);
