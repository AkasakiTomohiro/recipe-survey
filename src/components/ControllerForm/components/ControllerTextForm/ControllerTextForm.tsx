/**
 * @file ControllerTextForm.tsx
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import { useCallback } from 'react';
import { useController, useFormContext } from 'react-hook-form';

import { getZodErrorMessage } from '@/utils/ZodErrorParse';
import TextField from '@mui/material/TextField';

import { GridFormControl } from '../GridFormControl';
import { IControllerTextFormProps } from './types';

/**
 * Primary UI component for user interaction
 *
 * @example
 *
 * 利用したいコンポーネントでジェネリックに型を定義する形でラップして利用すること
 *
 * const TextForm = (props: IControllerTextFormProps<XXXType>) => {
 *   return (<ControllerTextForm {...props}/>);
 * };
 */
export const ControllerTextForm = ({
  name,
  isAllGird,
  onChange,
  inputErrorHelperText,
  isReadOnly,
  ...otherProps
}: IControllerTextFormProps): JSX.Element => {
  const { control, formState: { errors }} = useFormContext();
  const { field } = useController({ name: name, control });

  const errorMessage = getZodErrorMessage(name, errors);

  const onChangeWarp = useCallback((data: React.ChangeEvent<HTMLInputElement>): void => {
    let result: any = data.target.value;
    if(otherProps.type === 'number') {
      result = Number(data.target.value);
    }
    field.onChange({ target: { value: result }});
    if(onChange) {
      onChange(result);
    }
  }, [otherProps.type, field, onChange]);

  return (
    <GridFormControl isAllGird={isAllGird}>
      <TextField
        {...otherProps}
        {...field}
        onChange={onChangeWarp}
        error={Boolean(errorMessage)}
        helperText={Boolean(errorMessage) ? inputErrorHelperText : otherProps.helperText}
        variant="standard"
        inputProps={{ ...(otherProps.inputProps ?? {}), readOnly: isReadOnly }}
      />
    </GridFormControl>
  );
};
