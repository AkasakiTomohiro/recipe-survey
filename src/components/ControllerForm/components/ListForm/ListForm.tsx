/**
 * @file ListForm.tsx
 * @author Artan's Projects
 * @copyright Â© 2024 Artan's Projects. All rights reserved.
 */

import { useTranslations } from 'next-intl';
import { useCallback } from 'react';
import { useFieldArray, useFormContext } from 'react-hook-form';

import { DeleteIcon } from '@/components/Icon/Icon';
import { Box, Button, IconButton } from '@mui/material';
import MuiTable from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableRow from '@mui/material/TableRow';

import { FormRowProps, IListFormProps } from './types';

const FormRow = (props: FormRowProps): JSX.Element => {

  return (
    <TableRow hover key={`table-row-${props.index}`}>
      {props.forms(props.index).map((m, i) => <TableCell key={`table-row-${props.index}-cell-${i}`}>{m}</TableCell>)}
      {props.editable !== false &&
        <TableCell key={`table-row-${props.index}-delete-button`} sx={{ width: '16px' }}>
          <IconButton
            aria-label="delete"
            onClick={() => props.onClickDelete(props.index)}
            disabled={props.length === 1}
          >
            <DeleteIcon />
          </IconButton>
        </TableCell>
      }
    </TableRow>
  );
};

/**
 * Primary UI component for user interaction
 */
export const ListForm = (props: IListFormProps): JSX.Element => {
  const { control } = useFormContext();
  const t = useTranslations();

  const { fields, append, remove } = useFieldArray({
    control,
    name: props.name
  });

  const addRow = useCallback(() => {
    append(props.initialValue);
  }, [append, props.initialValue]);

  return (
    <>
      {
        props.editable && <Box sx={{
          width         : '100%',
          display       : 'flex',
          justifyContent: 'flex-end'
        }}>
          <Button
            variant="contained"
            onClick={addRow}
            disabled={fields.length >= props.maxLength}
          >{t('common.add')}</Button>
        </Box>
      }
      <TableContainer>
        <MuiTable sx={{ minWidth: 700 }} aria-label="list form">
          <TableBody>
            {fields.map((field, i) =>
              <FormRow
                key={field.id}
                index={i}
                forms={props.forms}
                onClickDelete={() => remove(i)}
                length={fields.length}
                editable={props.editable}
              />
            )}
          </TableBody>
        </MuiTable>
      </TableContainer>
    </>
  );
};
