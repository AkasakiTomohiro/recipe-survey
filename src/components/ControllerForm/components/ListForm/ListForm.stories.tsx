/* eslint-disable react-hooks/rules-of-hooks */
/**
 * @file ListForm.stories.tsx
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import exp from 'constants';
import { ComponentProps } from 'react';
import { FormProvider, useForm } from 'react-hook-form';

import { MaterialTypes, Units } from '@/lib/ProfileValidator';
import { expect } from '@storybook/jest';
import { userEvent, within } from '@storybook/testing-library';

import { ControllerTextForm } from '../ControllerTextForm';
import { ListForm } from './ListForm';

import type { Meta, StoryObj } from '@storybook/react';
type CustomProfile = MaterialTypes & Units

const meta: Meta<typeof ListForm> = {
  title : 'RecipeSurvey/components/ControllerForm/ListForm',
  render: (args) => {
    const methods = useForm<CustomProfile>({
      mode         : 'onChange',
      defaultValues: {
        units: [{
          unit       : '',
          description: ''
        }],
        materialTypes: [{
          materialType: ''
        }]
      }
    });
    return (
      <FormProvider {...methods} >
        <ListForm {...args} />
      </FormProvider>
    );
  }
};

export default meta;

type Story = StoryObj<typeof ListForm>;

export const Unit: Story = {
  args: {
    name : 'units',
    forms: (index: number): JSX.Element[] => {
      return [
        <ControllerTextForm label="エネルギー単位" name={`units.${index}.energyUnit`} key="energy_unit"/>,
        <ControllerTextForm label="説明" name={`units.${index}.description`} key="description" />
      ];
    },
    maxLength   : 4,
    initialValue: {
      energyUnit : '',
      description: ''
    }
  }
};

export const MaterialType: Story = {
  args: {
    name : 'materialTypes',
    forms: (index: number): JSX.Element[] => {
      return [
        <ControllerTextForm label="アイテム種別" name={`materialTypes.${index}.materialType`} key="material_type" />
      ];
    },
    maxLength   : 4,
    initialValue: {
      materialType: ''
    }
  }
};

export const ReadOnly: Story = {
  args: {
    name : 'materialTypes',
    forms: (index: number): JSX.Element[] => {
      return [
        <ControllerTextForm label="アイテム種別" name={`materialTypes.${index}.materialType`} key="material_type" />
      ];
    },
    maxLength   : 4,
    initialValue: {
      materialType: ''
    },
    editable: false
  }
};

const Template2 = (args: ComponentProps<typeof ListForm>) => {
  const methods = useForm<CustomProfile>({
    mode         : 'onChange',
    defaultValues: {
      units: [{
        unit       : '',
        description: ''
      }],
      materialTypes: [{
        materialType: 'M1'
      }]
    }
  });
  return (
    <FormProvider {...methods} >
      <ListForm {...args} />
    </FormProvider>
  );
}

export const AddRow: Story = {
  render: args => <Template2 {...args} />,
  args: {
    name : 'materialTypes',
    forms: (index: number): JSX.Element[] => {
      return [
        <ControllerTextForm label="アイテム種別" name={`materialTypes.${index}.materialType`} key="material_type" />
      ];
    },
    maxLength   : 4,
    initialValue: {
      materialType: ''
    },
    editable: true
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);

    // 初期値確認
    const button = canvas.getByRole('button', { name: /追加/i });
    await expect(button).toBeInTheDocument();
    const input1 = canvas.getByRole('textbox', { value: "M1" } as any);
    await expect(input1).toBeInTheDocument();

    // 追加ボタンを押下
    await userEvent.click(button);

    // フォームが増えているか確認
    const inputList = canvas.getAllByRole('textbox');
    expect(inputList).toHaveLength(2);
    const input2 = inputList[1];
    await expect(input2).toBeInTheDocument();
    await expect(input2).toHaveValue('');
    await userEvent.type(input2, 'M2');
    await expect(input2).toHaveValue('M2');
  }
}

export const DeleteRow: Story = {
  render: args => <Template2 {...args} />,
  args: {
    name : 'materialTypes',
    forms: (index: number): JSX.Element[] => {
      return [
        <ControllerTextForm label="アイテム種別" name={`materialTypes.${index}.materialType`} key="material_type" />
      ];
    },
    maxLength   : 4,
    initialValue: {
      materialType: ''
    },
    editable: true
  },
  play: async (context) => {
    const canvas = within(context.canvasElement);

    // 追加ボタンを押下のplayを実行
    await (AddRow as any).play(context);

    // M1の削除ボタンを押下
    const buttons = canvas.getAllByRole('button', {});
    expect(buttons).toHaveLength(3);
    const deleteButton = buttons[1];
    await expect(deleteButton).toBeInTheDocument();
    await userEvent.click(deleteButton);

    // M1が削除されているか確認
    const inputList = canvas.getAllByRole('textbox');
    expect(inputList).toHaveLength(1);
    const input1 = inputList[0];
    await expect(input1).toHaveValue('M2');
  }
}
