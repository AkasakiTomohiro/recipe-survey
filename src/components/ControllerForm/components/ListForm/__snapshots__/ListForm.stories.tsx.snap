// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeSurvey/components/ControllerForm/ListForm AddRow play-test 1`] = `
<div class="MuiBox-root css-2ertzu">
  <button class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-vbxad1-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
  >
    追加
    <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
      <span class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
            style="width: 56.3449px; height: 56.3449px; top: -10.1725px; left: 3.82754px;"
      >
        <span class="MuiTouchRipple-child MuiTouchRipple-childLeaving MuiTouchRipple-childPulsate">
        </span>
      </span>
      <span class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
            style="width: 77.1751px; height: 77.1751px; top: -20.5876px; left: -6.58756px;"
      >
        <span class="MuiTouchRipple-child MuiTouchRipple-childLeaving">
        </span>
      </span>
    </span>
  </button>
</div>
<div class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root">
  <table class="MuiTable-root css-twub5u-MuiTable-root"
         aria-label="list form"
  >
    <tbody class="MuiTableBody-root css-apqrd9-MuiTableBody-root">
      <tr class="MuiTableRow-root MuiTableRow-hover css-dvunen-MuiTableRow-root">
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-16blcct-MuiTableCell-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
            <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
              <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
                <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-zx681a-MuiFormLabel-root-MuiInputLabel-root"
                       data-shrink="true"
                       for=":r6:"
                       id=":r6:-label"
                >
                  アイテム種別
                </label>
                <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
                  <input aria-invalid="false"
                         id=":r6:"
                         name="materialTypes.0.materialType"
                         type="text"
                         class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                         value="M1"
                  >
                </div>
              </div>
            </div>
          </div>
        </td>
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-12g5br3-MuiTableCell-root">
          <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                  aria-label="delete"
          >
            <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                 focusable="false"
                 aria-hidden="true"
                 viewbox="0 0 24 24"
                 data-testid="DeleteIcon"
            >
              <path d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z">
              </path>
            </svg>
            <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
            </span>
          </button>
        </td>
      </tr>
      <tr class="MuiTableRow-root MuiTableRow-hover css-dvunen-MuiTableRow-root">
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-16blcct-MuiTableCell-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
            <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
              <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
                <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-zx681a-MuiFormLabel-root-MuiInputLabel-root"
                       data-shrink="true"
                       for=":r7:"
                       id=":r7:-label"
                >
                  アイテム種別
                </label>
                <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
                  <input aria-invalid="false"
                         id=":r7:"
                         name="materialTypes.1.materialType"
                         type="text"
                         class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                         value="M2"
                  >
                </div>
              </div>
            </div>
          </div>
        </td>
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-12g5br3-MuiTableCell-root">
          <button class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                  aria-label="delete"
          >
            <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                 focusable="false"
                 aria-hidden="true"
                 viewbox="0 0 24 24"
                 data-testid="DeleteIcon"
            >
              <path d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z">
              </path>
            </svg>
            <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
            </span>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`RecipeSurvey/components/ControllerForm/ListForm DeleteRow play-test 1`] = `
<div class="MuiBox-root css-2ertzu">
  <button class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-vbxad1-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
  >
    追加
    <span class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root">
      <span class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
            style="width: 56.3449px; height: 56.3449px; top: -10.1725px; left: 3.82754px;"
      >
        <span class="MuiTouchRipple-child MuiTouchRipple-childLeaving MuiTouchRipple-childPulsate">
        </span>
      </span>
      <span class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible"
            style="width: 77.1751px; height: 77.1751px; top: -20.5876px; left: -6.58756px;"
      >
        <span class="MuiTouchRipple-child MuiTouchRipple-childLeaving">
        </span>
      </span>
    </span>
  </button>
</div>
<div class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root">
  <table class="MuiTable-root css-twub5u-MuiTable-root"
         aria-label="list form"
  >
    <tbody class="MuiTableBody-root css-apqrd9-MuiTableBody-root">
      <tr class="MuiTableRow-root MuiTableRow-hover css-dvunen-MuiTableRow-root">
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-16blcct-MuiTableCell-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
            <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
              <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
                <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-zx681a-MuiFormLabel-root-MuiInputLabel-root"
                       data-shrink="true"
                       for=":ra:"
                       id=":ra:-label"
                >
                  アイテム種別
                </label>
                <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
                  <input aria-invalid="false"
                         id=":ra:"
                         name="materialTypes.0.materialType"
                         type="text"
                         class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                         value="M2"
                  >
                </div>
              </div>
            </div>
          </div>
        </td>
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-12g5br3-MuiTableCell-root">
          <button class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="-1"
                  type="button"
                  disabled
                  aria-label="delete"
          >
            <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                 focusable="false"
                 aria-hidden="true"
                 viewbox="0 0 24 24"
                 data-testid="DeleteIcon"
            >
              <path d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z">
              </path>
            </svg>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`RecipeSurvey/components/ControllerForm/ListForm MaterialType smoke-test 1`] = `
<div class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root">
  <table class="MuiTable-root css-twub5u-MuiTable-root"
         aria-label="list form"
  >
    <tbody class="MuiTableBody-root css-apqrd9-MuiTableBody-root">
      <tr class="MuiTableRow-root MuiTableRow-hover css-dvunen-MuiTableRow-root">
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-16blcct-MuiTableCell-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
            <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
              <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
                <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-11mm5xv-MuiFormLabel-root-MuiInputLabel-root"
                       data-shrink="false"
                       for=":r4:"
                       id=":r4:-label"
                >
                  アイテム種別
                </label>
                <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
                  <input aria-invalid="false"
                         id=":r4:"
                         name="materialTypes.0.materialType"
                         type="text"
                         class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                         value
                  >
                </div>
              </div>
            </div>
          </div>
        </td>
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-12g5br3-MuiTableCell-root">
          <button class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="-1"
                  type="button"
                  disabled
                  aria-label="delete"
          >
            <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                 focusable="false"
                 aria-hidden="true"
                 viewbox="0 0 24 24"
                 data-testid="DeleteIcon"
            >
              <path d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z">
              </path>
            </svg>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`RecipeSurvey/components/ControllerForm/ListForm ReadOnly smoke-test 1`] = `
<div class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root">
  <table class="MuiTable-root css-twub5u-MuiTable-root"
         aria-label="list form"
  >
    <tbody class="MuiTableBody-root css-apqrd9-MuiTableBody-root">
      <tr class="MuiTableRow-root MuiTableRow-hover css-dvunen-MuiTableRow-root">
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-16blcct-MuiTableCell-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
            <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
              <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
                <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-11mm5xv-MuiFormLabel-root-MuiInputLabel-root"
                       data-shrink="false"
                       for=":r5:"
                       id=":r5:-label"
                >
                  アイテム種別
                </label>
                <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
                  <input aria-invalid="false"
                         id=":r5:"
                         name="materialTypes.0.materialType"
                         type="text"
                         class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                         value
                  >
                </div>
              </div>
            </div>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`RecipeSurvey/components/ControllerForm/ListForm Unit smoke-test 1`] = `
<div class="MuiTableContainer-root css-rorn0c-MuiTableContainer-root">
  <table class="MuiTable-root css-twub5u-MuiTable-root"
         aria-label="list form"
  >
    <tbody class="MuiTableBody-root css-apqrd9-MuiTableBody-root">
      <tr class="MuiTableRow-root MuiTableRow-hover css-dvunen-MuiTableRow-root">
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-16blcct-MuiTableCell-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
            <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
              <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
                <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-11mm5xv-MuiFormLabel-root-MuiInputLabel-root"
                       data-shrink="false"
                       for=":r2:"
                       id=":r2:-label"
                >
                  エネルギー単位
                </label>
                <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
                  <input aria-invalid="false"
                         id=":r2:"
                         name="units.0.energyUnit"
                         type="text"
                         class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                         value
                  >
                </div>
              </div>
            </div>
          </div>
        </td>
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-16blcct-MuiTableCell-root">
          <div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
            <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
              <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
                <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-11mm5xv-MuiFormLabel-root-MuiInputLabel-root"
                       data-shrink="false"
                       for=":r3:"
                       id=":r3:-label"
                >
                  説明
                </label>
                <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
                  <input aria-invalid="false"
                         id=":r3:"
                         name="units.0.description"
                         type="text"
                         class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                         value
                  >
                </div>
              </div>
            </div>
          </div>
        </td>
        <td class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-12g5br3-MuiTableCell-root">
          <button class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="-1"
                  type="button"
                  disabled
                  aria-label="delete"
          >
            <svg class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                 focusable="false"
                 aria-hidden="true"
                 viewbox="0 0 24 24"
                 data-testid="DeleteIcon"
            >
              <path d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6zM19 4h-3.5l-1-1h-5l-1 1H5v2h14z">
              </path>
            </svg>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
